Update README and add a gemspec

digitalronin opened the pull request:
Make this project a gem, which simply installs
danger and any linter gems we specify into the
target project.

Update the README to contain full instructions for
adding Danger with the MoJ styles to an existing
project.

Commit:
Update README with instructions for adding Danger

Update the README to contain full instructions for
adding Danger with the Ministry of Justice styles
to an existing project.

SteveMarshall commented on README.md:
```diff
@@ -4,33 +4,39 @@
-### Configure Danger to use this repo's Dangerfile
-
-Commit a `Dangerfile` containing linking to this repo:
+1. Add the following to your Gemfile;
```
This now assumes the user is running Ruby, which isn't going to be true for a decent number of projects. It'd be good do break the instructions up into everyone, Ruby, and not-Ruby, maybe?

SteveMarshall commented on README.md:
```diff
@@ -4,33 +4,39 @@
+1. `bundle install`
 
-Add `danger` to your `Gemfile`.
+1. Create Dangerfile
```
Maybe “Create a `Dangerfile` that imports from the central one”?

SteveMarshall commented on README.md:
````diff
@@ -4,33 +4,39 @@
-```ruby
-gem 'danger'
+```bash
+echo "danger.import_dangerfile(github: 'ministryofjustice/danger')" > Dangerfile
````
😎

SteveMarshall commented on README.md:
```diff
@@ -4,33 +4,39 @@
 #### Run Danger in Travis
 
-Add Danger to `.travis.yml`:
+1. Add Danger to your .travis.yml (use whatever ruby version (>= 2.3.0) is appropriate)
```
`.travis.yml` should be in backticks.

SteveMarshall commented on README.md:
```diff
@@ -39,23 +45,34 @@ before_script:
+
+Add the environment variable `DANGER_GITHUB_API_TOKEN`, with the value
+from the password field of the `MOJ Danger bot Github API token` entry
+in [Rattic](https://rattic.service.dsd.io/cred/detail/695/)
```
Rattic's IP-restricted to our building, but I'm not sure we should necessarily be advertising its presence in our readmes? 🤔 Dunno. (Also, you could make the `MOJ Danger […] API token entry` the link text?)

@ministryofjustice/secops, any of you have thoughts either way?

SteveMarshall commented on README.md:
```diff
@@ -39,23 +45,34 @@ before_script:
+
+    https://travis-ci.org/ministryofjustice/[your project]/settings
+
+Add the environment variable `DANGER_GITHUB_API_TOKEN`, with the value
```
Is it worth being specific (or linking to documentation) about adding environment variables? Might be overkill, though.

SteveMarshall commented on README.md:
```diff
@@ -65,5 +82,15 @@ a Gemfile cluttering up your repo. In that case, you can (probably) use
+
+This [Github article](https://help.github.com/articles/changing-a-commit-message/)
+describes a process for doing this. But, please use `--force-with-lease` instead
+of `--force` when pushing your amendments. See [this article](https://developer.atlassian.com/blog/2015/04/force-with-lease/)
```
“use `--force-with-lease` instead of `--force`” might make more sense as the link text for the Atlassian article.

SteveMarshall commented on moj-danger.gemspec:
```diff
@@ -0,0 +1,22 @@
+# -*- encoding: utf-8 -*-
+
+Gem::Specification.new do |s|
+  s.name = 'moj-danger'
```
`moj-danger`… Any reason not to go with `ministryofjustice-danger`?

digitalronin replied:
Laziness. I'm happy to change it, if you like.

SteveMarshall replied:
Yeah, let's. We know what MOJ is, some random Rubygems user may not.

SteveMarshall commented on moj-danger.gemspec:
```diff
@@ -0,0 +1,22 @@
+Gem::Specification.new do |s|
+  s.name = 'moj-danger'
+  s.version = '0.1.0'
+  s.authors = ['Ministry Of Justice']
```
`of`, not `Of`

SteveMarshall commented on moj-danger.gemspec:
```diff
@@ -0,0 +1,22 @@
+  s.summary = 'Add Danger and linter gems'
+  s.homepage = 'https://github.com/ministryofjustice/danger'
+
+  s.files = %w(README.md LICENSE)
```
No `Dangerfile`?

digitalronin replied:
Adding the `Dangerfile` to the gem wouldn't do anything useful, IMO. It would end up in whatever location the gem got installed. You'd have to add a  command script to the gem to create/copy the `Dangerfile` to the local project (i.e. `moj-danger init`)

Since we give instructions in the README to create the appropriate Dangerfile, I didn't think it was useful to have it here.

SteveMarshall commented on Gemfile:
```diff
@@ -1,4 +0,0 @@
-source "https://rubygems.org"
```
Does bundler just use the gem spec without a Gemfile? Or should we keep this, but just have `source` and `gemspec`?

SteveMarshall replied:
In fact, the Travis build fails without a Gemfile, so we *do* need this, but it should point to `gemspec`, not explicitly list the gems.

SteveMarshall reviewed (changes requested):
I'd like to see this broken into a couple of commits, ideally one for the conversion to a gem and another (or a few?) for the documentation changes.

I've also added a few bits and bobs through the rest of the code.

Commit:
Clarify setup instructions for non-ruby projects

Also, fix readme point numbering - I thought Markdown did this
automatically, but it doesn't

Commit:
Remove Rattic URL

Convert to a gem: 'ministryofjustice-danger'

This installs danger and any linters required to implement
the Ministry of Justice house style rules for PRs.

SteveMarshall commented:
So, the sections in the readme are:

- `#` Danger (the top-level heading)
  - `##` Add Danger to your [ministryofjustice organization](https://github.com/ministryofjustice) Ruby project
  - `##` Add Danger to your [ministryofjustice organization](https://github.com/ministryofjustice) Non-ruby project
    - `####` Run Danger in Travis
      - `#####` If you're not running Ruby
    - `####` Run Danger in CircleCI
  - `##` Fixing your commits

That seems slightly odd to me, with “not Ruby” being under “in Travis” under “Non-ruby”. Perhaps they should instead be:

- `#` Danger (with some extra explanation of Danger being a tool for checking what we push to central repos, making running locally not as important?)
  - `##` Setup
    - `###` Add a Dangerfile to your project
    - `###` Set up Danger in your continuous integration environment
      - `####` If you're running Ruby
      - `####` If you're not running Ruby
      - `####` Travis
      - `####` CircleCI
  - `##` What we're checking (a new section, with a brief overview of the style rules?)
  - `##` (or maybe `###`?) Fixing your commits

I'd suggest this PR be about getting it to run elsewhere, and documenting that, and we spin out those last two headings I'm proposing (“what we're checking” and “fixing your commits”) into a separate PR. Waddya think?

Commit:
Remove deprecated moj-danger.gemspec

The gem is ministryofjustice-danger, now.

digitalronin commented:
Closing this PR, because I'm submitting 2 other PRs, based on the comments on this one.
